In [1]:
# importing libraries
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
# importing data set
data= pd.read_csv('student-data.csv')

In [3]:
# checking the data set

In [4]:
data.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  ...  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher  ...   
1     GP   F   17       U     GT3       T     1     1  at_home     other  ...   
2     GP   F   15       U     LE3       T     1     1  at_home     other  ...   
3     GP   F   15       U     GT3       T     4     2   health  services  ...   
4     GP   F   16       U     GT3       T     3     3    other     other  ...   

  internet romantic  famrel  freetime  goout Dalc Walc health absences passed  
0       no       no       4         3      4    1    1      3        6     no  
1      yes       no       5         3      3    1    1      3        4     no  
2      yes       no       4         3      2    2    3      3       10    yes  
3      yes      yes       3         2      2    1    1      5        2    yes  
4       no       no       4         3      2    1    2      5        4    yes  

[5 rows x 31 columns]

In [5]:
data.shape

(395, 31)

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 31 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   school      395 non-null    object
 1   sex         395 non-null    object
 2   age         395 non-null    int64 
 3   address     395 non-null    object
 4   famsize     395 non-null    object
 5   Pstatus     395 non-null    object
 6   Medu        395 non-null    int64 
 7   Fedu        395 non-null    int64 
 8   Mjob        395 non-null    object
 9   Fjob        395 non-null    object
 10  reason      395 non-null    object
 11  guardian    395 non-null    object
 12  traveltime  395 non-null    int64 
 13  studytime   395 non-null    int64 
 14  failures    395 non-null    int64 
 15  schoolsup   395 non-null    object
 16  famsup      395 non-null    object
 17  paid        395 non-null    object
 18  activities  395 non-null    object
 19  nursery     395 non-null    object
 20  higher    

In [7]:
data.isna().sum()

school        0
sex           0
age           0
address       0
famsize       0
Pstatus       0
Medu          0
Fedu          0
Mjob          0
Fjob          0
reason        0
guardian      0
traveltime    0
studytime     0
failures      0
schoolsup     0
famsup        0
paid          0
activities    0
nursery       0
higher        0
internet      0
romantic      0
famrel        0
freetime      0
goout         0
Dalc          0
Walc          0
health        0
absences      0
passed        0
dtype: int64

Data set is having 31 features and 395 observations. Ther is no null data present.Both categorical and numerical features are there. 

In [8]:
# checking the unique values present in categorical columns
for i in data.select_dtypes(include=['object']).columns:
     if i !='passed':
        print(i,np.unique(data[i]))
        continue

school ['GP' 'MS']
sex ['F' 'M']
address ['R' 'U']
famsize ['GT3' 'LE3']
Pstatus ['A' 'T']
Mjob ['at_home' 'health' 'other' 'services' 'teacher']
Fjob ['at_home' 'health' 'other' 'services' 'teacher']
reason ['course' 'home' 'other' 'reputation']
guardian ['father' 'mother' 'other']
schoolsup ['no' 'yes']
famsup ['no' 'yes']
paid ['no' 'yes']
activities ['no' 'yes']
nursery ['no' 'yes']
higher ['no' 'yes']
internet ['no' 'yes']
romantic ['no' 'yes']


We need to encode these values before checking correlation



# Label Encoding


In [9]:
#importing label encoder from skleran library
from sklearn.preprocessing import LabelEncoder
#creating a label encoder instance
Label_enc=LabelEncoder()
for i in data.select_dtypes(include=['object']).columns:
    if i!='passed':
        data[i]=Label_enc.fit_transform(data[i])
        continue
    

In [10]:
data.head()

school  sex  age  address  famsize  Pstatus  Medu  Fedu  Mjob  Fjob  ...  \
0       0    0   18        1        0        0     4     4     0     4  ...   
1       0    0   17        1        0        1     1     1     0     2  ...   
2       0    0   15        1        1        1     1     1     0     2  ...   
3       0    0   15        1        0        1     4     2     1     3  ...   
4       0    0   16        1        0        1     3     3     2     2  ...   

   internet  romantic  famrel  freetime  goout  Dalc  Walc  health  absences  \
0         0         0       4         3      4     1     1       3         6   
1         1         0       5         3      3     1     1       3         4   
2         1         0       4         3      2     2     3       3        10   
3         1         1       3         2      2     1     1       5         2   
4         0         0       4         3      2     1     2       5         4   

   passed  
0      no  
1      no  
2     yes  
3     yes  
4     yes  

[5 rows x 31 columns]

These columns are label encoded.

# one hot encoding 
school,sex,address,famsize,Pstatus,schoolsup,famsup,paid,activities,nursery,higher,internet,romantic columns


In [11]:
data=pd.get_dummies(data=data,columns=['school','sex','address','famsize','Pstatus','schoolsup','famsup',
               'paid','activities','nursery','higher','internet','romantic'])

In [12]:
data.head()

age  Medu  Fedu  Mjob  Fjob  reason  guardian  traveltime  studytime  \
0   18     4     4     0     4       0         1           2          2   
1   17     1     1     0     2       0         0           1          2   
2   15     1     1     0     2       2         1           1          2   
3   15     4     2     1     3       1         1           1          3   
4   16     3     3     2     2       1         0           1          2   

   failures  ...  activities_0  activities_1  nursery_0  nursery_1  higher_0  \
0         0  ...             1             0          0          1         0   
1         0  ...             1             0          1          0         0   
2         3  ...             1             0          0          1         0   
3         0  ...             0             1          0          1         0   
4         0  ...             1             0          0          1         0   

   higher_1  internet_0 internet_1  romantic_0  romantic_1  
0         1           1          0           1           0  
1         1           0          1           1           0  
2         1           0          1           1           0  
3         1           0          1           0           1  
4         1           1          0           1           0  

[5 rows x 44 columns]

In [13]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 44 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   age           395 non-null    int64 
 1   Medu          395 non-null    int64 
 2   Fedu          395 non-null    int64 
 3   Mjob          395 non-null    int64 
 4   Fjob          395 non-null    int64 
 5   reason        395 non-null    int64 
 6   guardian      395 non-null    int64 
 7   traveltime    395 non-null    int64 
 8   studytime     395 non-null    int64 
 9   failures      395 non-null    int64 
 10  famrel        395 non-null    int64 
 11  freetime      395 non-null    int64 
 12  goout         395 non-null    int64 
 13  Dalc          395 non-null    int64 
 14  Walc          395 non-null    int64 
 15  health        395 non-null    int64 
 16  absences      395 non-null    int64 
 17  passed        395 non-null    object
 18  school_0      395 non-null    uint8 
 19  school_1

In [14]:
# all categorical variables are converted to numerical.

# Splitting the data to train the model, Passed varibale is taken as the target variable

In [15]:
X = data.drop(['passed'], axis=1)
y = data['passed']
#importing train_test_split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 42, test_size = 0.2)

In [16]:
#Fitting the model SVM with Gaussian Kernel
#importing SVM
from sklearn.svm import SVC

In [17]:
svm_gauss=SVC(kernel='rbf')
svm_gauss.fit(X_train,y_train)
y_pred=svm_gauss.predict(X_test)

In [18]:
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report
print('Accuracy Score:',accuracy_score(y_test,y_pred))
print('Conclusion matrix\n',confusion_matrix(y_test,y_pred))

Accuracy Score: 0.6582278481012658
Conclusion matrix
 [[ 0 27]
 [ 0 52]]


# Random Forest model fitting

In [19]:
from sklearn.ensemble import RandomForestClassifier
rf_mod = RandomForestClassifier(random_state=42)
rf_mod.fit(X_train,y_train)
y_pred=rf_mod.predict(X_test)
print('Accuracy Score:',accuracy_score(y_test,y_pred))
print('Conclusion matrix\n',confusion_matrix(y_test,y_pred))

Accuracy Score: 0.6962025316455697
Conclusion matrix
 [[ 7 20]
 [ 4 48]]


SVM model is giving an accuracy of 65.8% where as random forest is giving 72% accuracy

In [20]:
# trying standard scaling
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()
X_train=scaler.fit_transform(X_train)
X_test=scaler.fit_transform(X_test)

In [21]:
#Random forest model after standard scaling
rf_mod.fit(X_train,y_train)
y_pred=rf_mod.predict(X_test)
print('Accuracy Score:',accuracy_score(y_test,y_pred))
print('Condusion matrix\n',confusion_matrix(y_test,y_pred))

Accuracy Score: 0.7341772151898734
Condusion matrix
 [[10 17]
 [ 4 48]]


In [22]:
#SVM model after standard scaling
svm_gauss.fit(X_train,y_train)
y_pred=svm_gauss.predict(X_test)
print('Accuracy Score:',accuracy_score(y_test,y_pred))
print('Conclusion matrix\n',confusion_matrix(y_test,y_pred))

Accuracy Score: 0.6708860759493671
Conclusion matrix
 [[ 4 23]
 [ 3 49]]


After standard scaling we can see there is a slight improvement in accuray of both models. But random forest is giving better accuracy at 73.14%